### PR TITLE
Add py3-dbs3-client && py3-dbs3-pycurl-client

### PR DESCRIPTION
**Pull request by todor-ivanov**
Do not merge yet!

### CODE REVIEW

**Comment by muhammadimranfarooqi** [comment]
+1 to c1268192b8b5155a997a93f64b787134c9427e80 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_c126819-20210423-165003.log" target="_blank">here</a>.

File: py3-dbs3-client.spec [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+## INITENV +PATH PYTHONPATH %i/${PYTHON_LIB_SITE_PACKAGES}
+## INITENV +PATH PYTHONPATH %i/x${PYTHON_LIB_SITE_PACKAGES}
+## INITENV SET DBS3_CLIENT_ROOT %i/
+## INITENV SET DBS3_CLIENT_VERSION %v
+## INITENV ALIAS dbs python $DBS3_CLIENT_ROOT/bin/dbs.py
+
+%define webdoc_files %{installroot}/%{pkgrel}/doc/
+Source0: git://github.com/dmwm/DBS.git?obj=master/%{realversion}&export=DBS&output=/%{n}.tar.gz
+Requires: python3 py3-cjson py3-dbs3-pycurl-client
+BuildRequires: py2-sphinx
```
Comment (amaltaro) [inline py3-dbs3-client.spec]: I think we should rely on a python3 sphinx spec.

File: py3-dbs3-client.spec [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+%setup -D -T -b 0 -n DBS
+
+%build
+python3 setup.py build_system -s dbs-client
+%install
+mkdir -p %i/etc/profile.d %i/{x,}{bin,lib,data,doc} %i/{x,}$PYTHON_LIB_SITE_PACKAGES
+python3 setup.py install_system -s dbs-client --prefix=%i
+find %i -name '*.egg-info' -exec rm {} \;
+
+# Generate dependencies-setup.{sh,csh} so init.{sh,csh} picks full environment.
```
Comment (amaltaro) [inline py3-dbs3-client.spec]: This block to generate dependencies are no longer needed. pkgtools does it for us (same comment on the other spec).

File: py3-dbs3-pycurl-client.spec [hunk 3]
```diff
@@ -0,0 +1,38 @@
+### RPM cms py3-dbs3-pycurl-client 3.16.0
+## INITENV +PATH PYTHONPATH %i/${PYTHON_LIB_SITE_PACKAGES}
+## INITENV +PATH PYTHONPATH %i/x${PYTHON_LIB_SITE_PACKAGES}
+
+#%define webdoc_files %{installroot}/%{pkgrel}/doc/
+Source0: git://github.com/dmwm/DBS.git?obj=master/%{realversion}&export=DBS&output=/%{n}.tar.gz
+Requires: python3 py3-cjson py3-pycurl curl
```
Comment (amaltaro) [inline py3-dbs3-pycurl-client.spec]: As you noticed, there is no cjson in python3, neither a py3-cjson spec file.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Todor, you need to add these packages as dependencies on the comp.spec file, otherwise we actually do not build them because there is nothing requesting them. Example:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/comp.spec#L21

**Comment by amaltaro** [comment]
@todor-ivanov Todor, I just had a chat with Yuyi and she said we could try updating the current specs to python3, instead of creating new ones. If we have issues with the codebase, we could try a `2to3` on a couple of directories in the DBS repository and see whether it brings us to a working dbs3-client / dbs3-pycurl-client.

**Comment by amaltaro** [comment]
Todor, I was actually wrong. We cannot use the same specs because there are many other services depending on those packages too. So creating new packages, as you did, is the way to go at the moment.

**Comment by todor-ivanov** [comment]
Thanks @amaltaro! Yes indeed I did check few `.spec` files before I proceeded, and actually keeping both  py2 && py3 spec files around for a while is not that much of a burden after all. BTW I am still having some troubles building them [1], even after getting rid of the `cjson` dependency and relying only on the automatic fallback mechanism as discussed during the Monday meeting. 

[1]
```
+ python3 setup.py build_system -s pycurl-client
running build_system
Traceback (most recent call last):
  File "setup.py", line 329, in <module>
    setup(name = 'dbs',
  File "/build/tivanov/w/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/build/tivanov/w/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/build/tivanov/w/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/distutils/dist.py", line 984, in run_command
    cmd_obj.ensure_finalized()
  File "/build/tivanov/w/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/distutils/cmd.py", line 107, in ensure_finalized
    self.finalize_options()
  File "setup.py", line 248, in finalize_options
    define_the_build(self, self.distribution, self.system, True, '')
  File "setup.py", line 206, in define_the_build
    py_version = (string.split(sys.version))[0]
AttributeError: module 'string' has no attribute 'split'
error: Bad exit status from /build/tivanov/w/tmp/rpm-tmp.TIyLD3 (%build)


RPM build errors:
    Macro %rpmbuild_libdir defined but not used within scope
    Bad exit status from /build/tivanov/w/tmp/rpm-tmp.TIyLD3 (%build)
```

**Comment by muhammadimranfarooqi** [comment]
+1 to 86ba5ea013a7aad2c9420670039890cafb18a922 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_86ba5ea-20210428-111004.log" target="_blank">here</a>.

File: py3-dbs3-pycurl-client.spec [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+BuildRequires: py2-sphinx
```
Comment (amaltaro) [inline py3-dbs3-pycurl-client.spec]: This we can update to `py3-sphinx` but in the end leave this line commented out.

File: py3-dbs3-client.spec (hunk 1 above)
Comment (todor-ivanov) [reply py3-dbs3-client.spec]: fixed now

File: py3-dbs3-client.spec (hunk 2 above)
Comment (todor-ivanov) [reply py3-dbs3-client.spec]: fixed

File: py3-dbs3-pycurl-client.spec (hunk 3 above)
Comment (todor-ivanov) [reply py3-dbs3-pycurl-client.spec]: fixed

File: py3-dbs3-pycurl-client.spec (hunk 4 above)
Comment (todor-ivanov) [reply py3-dbs3-pycurl-client.spec]: fixed

**Comment by muhammadimranfarooqi** [comment]
+1 to f558e7774f9f998d3732739dc5a74f7b8d9850bd on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_f558e77-20210512-140503.log" target="_blank">here</a>.

File: py3-dbs3-pycurl-client.spec (hunk 4 above)
Comment (todor-ivanov) [reply py3-dbs3-pycurl-client.spec]: fixed

File: py3-dbs3-pycurl-client.spec (hunk 3 above)
Comment (todor-ivanov) [reply py3-dbs3-pycurl-client.spec]: fixed

**Comment by amaltaro** [comment]
Thanks for providing these changes, Todor.

Before we forget, this PR cannot be merged until we define `py3-dbs3-client` as a dependency in the `wmagentpy3.spec` file, replacing the python2 based client in this line:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/wmagentpy3.spec#L8

**Comment by muhammadimranfarooqi** [comment]
+1 to 2ff5ed84cf21d2e5aa3160452e962d242ccc67f3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_2ff5ed8-20210513-182003.log" target="_blank">here</a>.

**Comment by todor-ivanov** [comment]
Hi @amaltaro , Sorry if I misunderstood something, but I actually do not see a reason why it couldn't be merged before the dependency is put in place at the `wmagentpy3.spec` file. Of course it is a must to be done so that the work with the rest of the py3 environment changes may continue, but I do not see a constrain about the order in which these are to be merged.

**Comment by todor-ivanov** [comment]
Thinking twice - actually may be right @amaltaro . Not that it is impossible, but definitely no need for a separate PR  just for adding the dependency to `wmagentpy3.spec`. It is fixed in the current one now.

**Comment by muhammadimranfarooqi** [comment]
-1 to 619ea351588228bcffa8f32873fc63e0b74f472b on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_619ea35-20210514-123004.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
If you provide only those two new specs and nothing else, the build bot isn't really going to test them. So you need to have a last piece which will trigger the building process for those 2 specs. Adding the dependency to the wmagentpy3 spec is one way to get it properly (and you can see with the -1 in the result ;))

**Comment by muhammadimranfarooqi** [comment]
+1 to 2f2ae1b1083f6743781712cf831032ae88ddb5ec on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/6837/test_2f2ae1b-20210518-134003.log" target="_blank">here</a>.

**Pull request merged**